pkg/retriever/bitswaphelpers/multiblockstore.go

```diff
@@ -0,0 +1,157 @@
+package bitswaphelpers
+
+import (
+	"context"
+	"errors"
+	"io"
+	"sync"
+
+	"github.com/filecoin-project/lassie/pkg/types"
+	"github.com/ipfs/go-cid"
+	format "github.com/ipfs/go-ipld-format"
+	"github.com/ipfs/go-libipfs/blocks"
+	"github.com/ipld/go-ipld-prime/linking"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+)
+
+// ErrNotSupported indicates an operation not supported by the MultiBlockstore
+var ErrNotSupported = errors.New("not supported")
+
+// ErrAlreadyRegistered means something has already been registered for a retrieval id
+var ErrAlreadyRegisterd = errors.New("already registered")
+
+// ErrAlreadyRegistered means there is nothing registered for a retrieval id
+var ErrNotRegistered = errors.New("not registered")
+
+type linkSystemRecord struct {
+	lsys   *linking.LinkSystem
+	cancel context.CancelFunc
+}
+
+// MultiBlockstore creates a blockstore based on one or more linkystems, extracting the target linksystem for each request
+// from the retrieval id context key
+type MultiBlockstore struct {
+	linkSystems   map[types.RetrievalID]*linking.LinkSystem
+	linkSystemsLk sync.RWMutex
+}
+
+// NewMultiblockstore returns a new MultiBlockstore
+func NewMultiblockstore() *MultiBlockstore {
+	return &MultiBlockstore{
+		linkSystems: make(map[types.RetrievalID]*linking.LinkSystem),
+	}
+}
+
+// AddLinkSystem registers a linksystem to use for a given retrieval id
+func (mbs *MultiBlockstore) AddLinkSystem(id types.RetrievalID, lsys *linking.LinkSystem) error {
+	mbs.linkSystemsLk.Lock()
+	defer mbs.linkSystemsLk.Unlock()
+	if _, ok := mbs.linkSystems[id]; ok {
+		return ErrAlreadyRegisterd
+	}
+	mbs.linkSystems[id] = lsys
+	return nil
+}
+
+// RemoveLinkSystem unregisters the link system for a given retrieval id
+func (mbs *MultiBlockstore) RemoveLinkSystem(id types.RetrievalID) {
+	mbs.linkSystemsLk.Lock()
+	defer mbs.linkSystemsLk.Unlock()
+	delete(mbs.linkSystems, id)
+}
+
+// DeleteBlock is not supported
+func (mbs *MultiBlockstore) DeleteBlock(ctx context.Context, c cid.Cid) error {
+	return ErrNotSupported
+}
+
+// Has is not supported
+func (mbs *MultiBlockstore) Has(ctx context.Context, c cid.Cid) (bool, error) {
+	return false, ErrNotSupported
+}
+
+type byteReader interface {
+	Bytes() []byte
+}
+
+// Get returns a block only if the given ctx contains a retrieval ID as a value that
+// references a known linksystem. If it does, it uses that linksystem to load the block
+func (mbs *MultiBlockstore) Get(ctx context.Context, c cid.Cid) (blocks.Block, error) {
+	id, err := types.RetrievalIDFromContext(ctx)
+	if err != nil {
+		if errors.Is(err, types.ErrMissingContextKey) {
+			return nil, format.ErrNotFound{Cid: c}
+		}
```

Comment:
Shouldn't this indicate we've screwed up? If we get to this point then ISTM that we've prematurely removed the linksystem for this retrieval. Or do we expect Bitswap to be doing some extra work for a session beyond the bounds of what we expect from a single traversal?

Reply:
truthfully I think we should never hit this. I guess actually it may make sense not to correct the error cause if we get here, something went wrong